File: python_solution/award.py
class Award(object):
    def __init__(self, name=None, expires_in=None, quality=None):
        self.name = name
        self.expires_in = expires_in
        self.quality = quality

    def _valid_quality(self):
        return self.quality in range(1, 50)

    def _increase_quality_by(self, value):
        self.quality += value

    def _decrease_quality_by(self, value):
        self.quality -= value

    def quality_before_expiration(self):
        if self._valid_quality():
            if self.name == "Blue First":
                self._increase_quality_by(1)
            elif self.name == "Blue Compare":
                if self.expires_in <= 10:
                    self._increase_quality_by(3 if self.expires_in <= 5 else 2)
                else:
                    self._increase_quality_by(1)
            elif self.name == "Blue Star":
                self._decrease_quality_by(2)
            else:
                self._decrease_quality_by(1)

    def quality_pass_expiration(self):
        self.expires_in -= 1
        if self.expires_in < 0 and self._valid_quality():
            if self.name == "Blue First":
                self._increase_quality_by(1)
            elif self.name == "Blue Compare":
                self._decrease_quality_by(self.quality)
            elif self.name == "Blue Star":
                self._decrease_quality_by(2)
            else:
                self._decrease_quality_by(1)

File: ruby_solution/award.rb
require "award_types/BlueFirst"
require "award_types/BlueCompare"
require "award_types/BlueStar"
require "award_types/NormalAward"

class Award
  attr_accessor :name, :expires_in, :quality

  def initialize(name, expires_in, quality)
    @name = name
    @expires_in = expires_in
    @quality = quality
  end

  def add_quality(value)
    self.quality += value if (self.quality + value).between?(0, 50)
  end

  def calculate_quality
    return if self.name == "Blue Distinction Plus"
    valid_product = {
      "Blue First": BlueFirst.new(self).current_value,
      "Blue Compare": BlueCompare.new(self).current_value,
      "Blue Star": BlueStar.new(self).current_value,
    }
    self.expires_in -= 1
    add_quality(valid_product[self.name.to_sym] ? valid_product[self.name.to_sym] : NormalAward.new(self).current_value)
  end
end
